File: simulators/typescript/src/scenarios/transactions.ts
import { Pool } from "pg";
import { Config, RawResult } from "../types";

export async function transactions(config: Config): Promise<RawResult> {
  const pool = new Pool({ host: config.host, port: config.port, user: config.user, password: config.password, database: config.database, max: config.concurrency });
  pool.on("error", () => {}); // suppress unhandled pool-level errors
  let ops = 0, errors = 0;
  const latencies: number[] = [];

  const start = Date.now();
  const workers = Array.from({ length: config.concurrency }, (_, i) => (async () => {
    for (let iter = 0; iter < 10; iter++) {
      // Commit flow
      let t = performance.now();
      const client = await pool.connect();
      try {
        await client.query("BEGIN");
        await client.query("INSERT INTO users (name, email, balance) VALUES ($1, $2, $3)", [`tx_user_${i}_${iter}`, `tx_commit_${i}_${iter}@test.com`, 500.0]);
        await client.query("COMMIT");
      } catch { errors++; try { await client.query("ROLLBACK"); } catch {} }
      finally { client.release(); }
      latencies.push(performance.now() - t); ops++;

      // Rollback flow
      t = performance.now();
      const client2 = await pool.connect();
      try {
        await client2.query("BEGIN");
        await client2.query("INSERT INTO users (name, email, balance) VALUES ($1, $2, $3)", ["will_rollback", `rb_${i}_${iter}@test.com`, 999.99]);
        await client2.query("ROLLBACK");
      } catch { try { await client2.query("ROLLBACK"); } catch {} }
      finally { client2.release(); }
      latencies.push(performance.now() - t); ops++;

      // Savepoint flow
      t = performance.now();
      const client3 = await pool.connect();
      try {
        await client3.query("BEGIN");
        await client3.query("SAVEPOINT sp1");
        await client3.query("INSERT INTO users (name, email, balance) VALUES ($1, $2, $3)", ["sp_user", `sp_${i}_${iter}@test.com`, 100.0]);
        await client3.query("ROLLBACK TO SAVEPOINT sp1");
        await client3.query("COMMIT");
      } catch { errors++; try { await client3.query("ROLLBACK"); } catch {} }
      finally { client3.release(); }
      latencies.push(performance.now() - t); ops++;
    }
  })());

  await Promise.all(workers);
  await pool.end();
  return { totalOps: ops, errors, durationMs: Date.now() - start, latencies };
}

File: simulators/go/scenarios/transactions.go
package scenarios

import (
	"context"
	"fmt"
	"math/rand/v2"
	"sync"
	"sync/atomic"
	"time"
)

// Transactions exercises BEGIN/COMMIT, BEGIN/ROLLBACK, and SAVEPOINT flows.
func Transactions(ctx context.Context, cfg *Config) (*Result, error) {
	// Unique per-run prefix prevents duplicate key violations on repeated runs.
	runID := rand.Uint64()

	var (
		ops       atomic.Int64
		errCount  atomic.Int64
		mu        sync.Mutex
		latencies []time.Duration
	)
	errs := NewErrorSampler(10)

	start := time.Now()
	var wg sync.WaitGroup

	for i := 0; i < cfg.Concurrency; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()

			conn, err := ConnectClient(ctx, cfg)
			if err != nil {
				ops.Add(1)
				errCount.Add(1)
				errs.Record(err)
				return
			}
			defer conn.Close(ctx)

			for iter := 0; iter < 10; iter++ {
				// --- Commit flow ---
				t := time.Now()
				tx, err := conn.Begin(ctx)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
					ops.Add(1)
					continue
				}

				email := fmt.Sprintf("tx_%d_%d_%d@test.com", runID, workerID, iter)
				_, err = tx.Exec(ctx, "INSERT INTO users (name, email, balance) VALUES ($1, $2, $3)",
					fmt.Sprintf("tx_user_%d_%d", workerID, iter), email, 500.00)
				if err != nil {
					_ = tx.Rollback(ctx)
					errCount.Add(1)
					errs.Record(err)
					ops.Add(1)
					continue
				}

				err = tx.Commit(ctx)
				d := time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
				}

				// --- Rollback flow ---
				t = time.Now()
				tx, err = conn.Begin(ctx)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
					ops.Add(1)
					continue
				}

				_, _ = tx.Exec(ctx, "INSERT INTO users (name, email, balance) VALUES ($1, $2, $3)",
					"will_rollback", fmt.Sprintf("rb_%d_%d@test.com", workerID, iter), 999.99)

				err = tx.Rollback(ctx)
				d = time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
				}

				// --- Savepoint flow ---
				t = time.Now()
				tx, err = conn.Begin(ctx)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
					ops.Add(1)
					continue
				}

				_, _ = tx.Exec(ctx, "SAVEPOINT sp1")
				spEmail := fmt.Sprintf("sp_%d_%d@test.com", workerID, iter)
				_, _ = tx.Exec(ctx, "INSERT INTO users (name, email, balance) VALUES ($1, $2, $3)",
					"savepoint_user", spEmail, 100.00)
				_, _ = tx.Exec(ctx, "ROLLBACK TO SAVEPOINT sp1")
				err = tx.Commit(ctx)
				d = time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
				}
			}
		}(i)
	}

	wg.Wait()
	duration := time.Since(start)

	return &Result{
		TotalOps:  int(ops.Load()),
		Errors:    int(errCount.Load()),
		Duration:  duration,
		Latencies: latencies,
		Notes:     errs.Notes(),
	}, nil
}
